ui/dmt_mappinsubjectmanager.lua

```diff
@@ -0,0 +1,301 @@
+    local allPins = {};
+    local mapPinSubjectTable = GetMapPinSubjectTable(playerID, MAIN_CACHE_KEY) or {};
+    for key, pinSubject in pairs(mapPinSubjectTable) do
+        table.insert(allPins, pinSubject);
+    end
+    return allPins;
+end
+
+-- Set/Update the MapPinSubject at the given position from the provided value.
+function UpdateMapPinSubject(playerID:number, posX:number, posY:number, pinSubject:table)
+    -- Get the saved MapPinSubject data for the plot
+    local mapPinSubjectTable = GetMapPinSubjectTable(playerID, MAIN_CACHE_KEY) or {};
+    local savedPinSubject = mapPinSubjectTable[GetMapPinSubjectCacheKey(posX, posY)];
+    -- Skip if there is no difference between the saved data and the new data
+    if DeepCompare(savedPinSubject, pinSubject) then
+        return;
+    end
+    -- Save the new MapPinSubject data
+    mapPinSubjectTable[GetMapPinSubjectCacheKey(posX, posY)] = pinSubject;
+    SaveMapPinSubjectTable(playerID, MAIN_CACHE_KEY, mapPinSubjectTable);
+end
+
+-- Get the cache key to be used for storing MapPinSubjects.
+function GetMapPinSubjectCacheKey(posX:number, posY:number)
+    return posX .. "_" .. posY;
+end
+
+-- Get table from player config or cache and deserialize.
+function GetMapPinSubjectTable(playerID:number, key:string)
+    -- Get the player configuration
+    local playerConfig = PlayerConfigurations[playerID];
+    if playerConfig == nil then
+        return nil;
+    end
+    -- Use the cache if it is active and has data saved
+    if bUseCache and mapPinSubjectCache[playerID] then
+        return mapPinSubjectCache[playerID];
+    end
+    -- Get the data from the player config and deserialize it if it exists
+    local serializedTable = playerConfig:GetValue(key)
+    if serializedTable ~= nil then
+        local deserializedTable = unpickle(serializedTable);
+        -- Save the data to the cache if it is active
+        if bUseCache then
+            mapPinSubjectCache[playerID] = deserializedTable;
+        end
+        -- Return the deserialized table
+        return deserializedTable;
+    end
+    -- If this point is reached, just return nil
+    return nil;
+end
+
+-- Serialize and save table to player config or cache.
+function SaveMapPinSubjectTable(playerID:number, key:string, value:table)
+    -- Get the player configuration
+    local playerConfig = PlayerConfigurations[playerID];
+    if playerConfig == nil then
+        return;
+    end
+    -- Save the data to the cache if it is active
+    if bUseCache then
+        mapPinSubjectCache[playerID] = value;
+        bSaveCache = true;
+        return;
+    end
+    -- Serialize and save the data to the player config
+    local serializedTable = pickle(value)
+    playerConfig:SetValue(key, serializedTable)
+    -- Mark as having unbroadcasted changes
+    bUnbroadcastedChanges = true;
+end
+
+-- Initialize the MapPinSubject cache for faster reading/writing
+function InitializeMapPinSubjectCache(bForceInitialize:boolean)
```

Comment:
Hey I went over all the caching logic here and have some thoughts for improvements. Seems like the cache is needed due to serialization/deserialization performance. 

One option is trying other serialization options. Like MetaLua used by "Endgame Map Replay" mod, or DataDumper used by "More Lenses" and "Better Trade Screen".

Another idea is to optimize the usage of the cache. I noticed you are calling init cache for methods like onPinAdded or updatePin, but not for other read methods like isAdjacentToXXX. Thinking about the values being cached, they will only be updated when it is explicitly updated by the current user. I wonder if we can do the following:
1. In GetMapPinSubjectTable, only read from player config if mapPinSubjectCache[playerID] is nil, and the cache is set after the first read from player config. So most reads should read from the cache directly, except the very first read.
2. In UpdateMapPinSubject, take one additional parameter "forceUpdate". All updates get stored to mapPinSubjectCache, and if forceUpdate is true, store to the player config and perform the broadcast. 
3. Add another UpdatePlayerConfig method that forces writing cached data to player config and perform the broadcast.
4. Then in methods like IsValidCityCenterPosition, set forceUpdate to true. But in methods that updates multiple pins at the same time, set forceUpdate to false, and at the end of the method or update, call UpdatePlayerConfig.

The cache is always updated in this case, at least for the current player. I'm not sure how does the current player get play config updates from other players. Will we get it through MapPinManager's Refresh method? If so, before calling BASE_Refresh, you can call a new invalidateCache method to invalid mapPinSubjectCache for all other player ids except the current player's, setting them to nil. So next reads to other players map pin will be reading from player config directly.

Let me know if there's any edge case that I haven't considered.

Reply:
Thanks for noting the serialization options used by those other mods! I was conflicted on which serialization method to use, and I settled on the one I originally had here before it was all I found when searching on the CivFanatics forum and Civ VI Modding Helpline Discord server.

I tried the ones you shared out and they seem to perform faster than the one I originally chose. I didn't notice any significant difference between MetaLua or DataDumper, so I just chose MetaLua since it seems simpler. As a result, I was able to remove the caching code as it's no longer needed for performance. This reduces the complexity of my added code too.

However, I'm happy to continue the cache discussion if we wish to maybe re-add it for some other reason. I have my responses to your comment below.

---
<details>
<summary>Optimizing the usage of the cache</summary><br>
Is it possible for those other read methods to be called without one of the event handlers being called first (in dmt_yieldcalculator)? I was thinking that it was only necessary to initialize the cache when the events are called because there was no way for the MapPinSubjects to be read/written without those being called first. If it is possible, we could definitely restructure the cache setup to be more like what you proposed. <br><br>
My original idea with the cache was for it to be as minimally invasive as possible. I wasn't sure how you would feel with an always active cache, so I made it so it would only turn on when map pins were about to be updated, and then it would shut down afterwards. If we bring back the cache though and you're okay with it, we could definitely set it up like you proposed.
</details>

---
<details>
<summary>How does the current player get config updates from other players?</summary><br>
To answer this question near the end, the local client gets the player config updates from the other clients through the `Network.BroadcastPlayerInfo()` command. When used, the game (somewhere behind the scenes) updates the player configuration data on every other client and then triggers the `Events.PlayerInfoChanged` event for everyone. In the original `MapPinManager.lua` file, the handler for this event is `OnPlayerInfoChanged()` which then simply calls the `Refresh()` function. So this is how each client gets the data from all other clients, including itself.<br><br>
This is also how I was able to remove the `RefreshMapPinUI()` function. Since the `Events.PlayerInfoChanged` is triggered even when the local player's config is updated, there was no need for that function since the event served the same purpose.
</details>

---
<details>
<summary>Using a new invalidateCache method to invalidate the cache for all other players except the current player</summary><br>
For your idea regarding the new invalidateCache method, if we brought back the cache, we can actually narrow this idea a bit more. The `Events.PlayerInfoChanged` event gives a Player ID when it's triggered, so we could just invalidate the cache for that specific player.<br><br>
My original idea here with the cache was to be minimally invasive, so it would only be active when it needed to be. So the cache is shut off until `Refresh()` was called because that was the primary function that updates all the pins on the UI. I thought that was the best place to put it so the cache wouldn't be active more than it had to be. After that function finished, the cache would shut back down in my code. <br><br>
However, if we expanded it so the cache was active pretty much at all times, then we could just use the event to invalidate the one player and keep the other players' data.
</details>

---